src/nvim/fileio.c

```diff
@@ -2736,156 +2736,172 @@ buf_write(
-                                       no_prepend_dot);
-          }
-
-          if (backup == NULL) {
-            xfree(rootname);
-            some_error = TRUE;                          /* out of memory */
-            goto nobackup;
-          }
-
-          /*
-           * Check if backup file already exists.
-           */
-          if (os_fileinfo((char *)backup, &file_info_new)) {
-            if (os_fileinfo_id_equal(&file_info_new, &file_info_old)) {
-              //
-              // Backup file is same as original file.
-              // May happen when modname() gave the same file back (e.g. silly
-              // link). If we don't check here, we either ruin the file when
-              // copying or erase it after writing.
-              //
-              XFREE_CLEAR(backup);              // no backup file to delete
-            } else if (!p_bk) {
-              // We are not going to keep the backup file, so don't
-              // delete an existing one, and try to use another name instead.
-              // Change one character, just before the extension.
-              //
-              wp = backup + STRLEN(backup) - 1 - STRLEN(backup_ext);
-              if (wp < backup) {                // empty file name ???
-                wp = backup;
-              }
-              *wp = 'z';
-              while (*wp > 'a'
-                     && os_fileinfo((char *)backup, &file_info_new)) {
-                --*wp;
-              }
-              // They all exist??? Must be something wrong.
-              if (*wp == 'a') {
-                XFREE_CLEAR(backup);
-              }
-            }
-          }
-        }
-        xfree(rootname);
-
-        /*
-         * Try to create the backup file
-         */
-        if (backup != NULL) {
-          /* remove old backup, if present */
-          os_remove((char *)backup);
-
-          // set file protection same as original file, but
-          // strip s-bit.
-          (void)os_setperm((const char *)backup, perm & 0777);
-
-#ifdef UNIX
-          //
-          // Try to set the group of the backup same as the original file. If
-          // this fails, set the protection bits for the group same as the
-          // protection bits for others.
-          //
-          if (file_info_new.stat.st_gid != file_info_old.stat.st_gid
-              && os_chown((char *)backup, -1, file_info_old.stat.st_gid) != 0) {
-            os_setperm((const char *)backup,
-                       (perm & 0707) | ((perm & 07) << 3));
-          }
-#endif
-
-          // copy the file
-          if (os_copy((char *)fname, (char *)backup, UV_FS_COPYFILE_FICLONE)
-              != 0) {
-            SET_ERRMSG(_("E506: Can't write to backup file "
-                         "(add ! to override)"));
-          }
-
-#ifdef UNIX
-          os_file_settime((char *)backup,
-                          file_info_old.stat.st_atim.tv_sec,
-                          file_info_old.stat.st_mtim.tv_sec);
-#endif
```

Comment:
@teto, A question:

Over here, if we fail copying the original file, we still try to set atime and mtime on the backup file. Is that intentional? What is the point of this if there is no file yet? If this isn't the case then I must have misunderstood something.

Reply:
It does seem weird. I tried a git blame on neovim source code but with little success. Checking against the vim source code (grep for buf_write), it seems to do the same (does not check against vim_rename success). This seems wrong.